File: backend/Controllers/AudioController.cs
using Microsoft.AspNetCore.Mvc;

namespace Backend.Controllers;

[ApiController]
[Route("api/[controller]")]
public class AudioController : ControllerBase
{
    private readonly string _audioDir;

    public AudioController(IWebHostEnvironment env)
    {
        _audioDir = Path.Combine(env.ContentRootPath, "AudioFiles");
        if (!Directory.Exists(_audioDir))
        {
            Directory.CreateDirectory(_audioDir);
        }
    }

    /// <summary>
    /// GET /api/audio/check/{vocabId}/{lang}
    /// Check if an audio file exists for a vocab/language combination.
    /// lang: "d" (German) or "i" (Indonesian)
    /// Returns { exists: true/false }
    /// </summary>
    [HttpGet("check/{vocabId}/{lang}")]
    public ActionResult CheckAudio(int vocabId, string lang)
    {
        if (lang != "d" && lang != "i")
        {
            return BadRequest(new { message = "Sprache muss 'd' oder 'i' sein / Bahasa harus 'd' atau 'i'" });
        }

        var fileName = $"{vocabId}{lang}.webm";
        var filePath = Path.Combine(_audioDir, fileName);
        return Ok(new { exists = System.IO.File.Exists(filePath) });
    }

    /// <summary>
    /// GET /api/audio/{vocabId}/{lang}
    /// Stream the audio file for playback.
    /// </summary>
    [HttpGet("{vocabId}/{lang}")]
    public ActionResult GetAudio(int vocabId, string lang)
    {
        if (lang != "d" && lang != "i")
        {
            return BadRequest(new { message = "Sprache muss 'd' oder 'i' sein / Bahasa harus 'd' atau 'i'" });
        }

        var fileName = $"{vocabId}{lang}.webm";
        var filePath = Path.Combine(_audioDir, fileName);

        if (!System.IO.File.Exists(filePath))
        {
            return NotFound(new { message = "Audiodatei nicht gefunden / File audio tidak ditemukan" });
        }

        var stream = System.IO.File.OpenRead(filePath);
        return File(stream, "audio/webm", fileName);
    }

    /// <summary>
    /// POST /api/audio/{vocabId}/{lang}
    /// Upload a new audio recording. Accepts multipart/form-data with an "audio" file field.
    /// Overwrites any existing file (for quality improvement).
    /// </summary>
    [HttpPost("{vocabId}/{lang}")]
    public async Task<ActionResult> UploadAudio(int vocabId, string lang, IFormFile audio)
    {
        if (lang != "d" && lang != "i")
        {
            return BadRequest(new { message = "Sprache muss 'd' oder 'i' sein / Bahasa harus 'd' atau 'i'" });
        }

        if (audio == null || audio.Length == 0)
        {
            return BadRequest(new { message = "Keine Audiodatei gesendet / Tidak ada file audio yang dikirim" });
        }

        // Limit file size to 5MB
        if (audio.Length > 5 * 1024 * 1024)
        {
            return BadRequest(new { message = "Datei zu groß (max 5MB) / File terlalu besar (maks 5MB)" });
        }

        var fileName = $"{vocabId}{lang}.webm";
        var filePath = Path.Combine(_audioDir, fileName);

        using var stream = new FileStream(filePath, FileMode.Create);
        await audio.CopyToAsync(stream);

        return Ok(new { message = "Audio gespeichert / Audio disimpan", fileName });
    }
}

File: backend-java/src/main/java/com/vocabtrainer/controller/AudioController.java
package com.vocabtrainer.controller;

import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

@RestController
@RequestMapping("/api/audio")
public class AudioController {

    private final Path audioDir;

    public AudioController() {
        this.audioDir = Paths.get(System.getProperty("user.dir"), "AudioFiles");
        try {
            Files.createDirectories(audioDir);
        } catch (IOException e) {
            throw new RuntimeException("Could not create AudioFiles directory", e);
        }
    }

    @GetMapping("/check/{vocabId}/{lang}")
    public ResponseEntity<?> checkAudio(@PathVariable int vocabId, @PathVariable String lang) {
        if (!"d".equals(lang) && !"i".equals(lang)) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Sprache muss 'd' oder 'i' sein / Bahasa harus 'd' atau 'i'"));
        }

        String fileName = vocabId + lang + ".webm";
        Path filePath = audioDir.resolve(fileName);
        return ResponseEntity.ok(Map.of("exists", Files.exists(filePath)));
    }

    @GetMapping("/{vocabId}/{lang}")
    public ResponseEntity<?> getAudio(@PathVariable int vocabId, @PathVariable String lang) {
        if (!"d".equals(lang) && !"i".equals(lang)) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Sprache muss 'd' oder 'i' sein / Bahasa harus 'd' atau 'i'"));
        }

        String fileName = vocabId + lang + ".webm";
        Path filePath = audioDir.resolve(fileName);

        if (!Files.exists(filePath)) {
            return ResponseEntity.status(404)
                    .body(Map.of("message", "Audiodatei nicht gefunden / File audio tidak ditemukan"));
        }

        Resource resource = new FileSystemResource(filePath);
        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType("audio/webm"))
                .header(HttpHeaders.CONTENT_DISPOSITION, "inline; filename=\"" + fileName + "\"")
                .body(resource);
    }

    @PostMapping("/{vocabId}/{lang}")
    public ResponseEntity<?> uploadAudio(
            @PathVariable int vocabId,
            @PathVariable String lang,
            @RequestParam("audio") MultipartFile audio) {

        if (!"d".equals(lang) && !"i".equals(lang)) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Sprache muss 'd' oder 'i' sein / Bahasa harus 'd' atau 'i'"));
        }

        if (audio.isEmpty()) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Keine Audiodatei gesendet / Tidak ada file audio yang dikirim"));
        }

        if (audio.getSize() > 5 * 1024 * 1024) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Datei zu groß (max 5MB) / File terlalu besar (maks 5MB)"));
        }

        String fileName = vocabId + lang + ".webm";
        Path filePath = audioDir.resolve(fileName);

        try {
            Files.write(filePath, audio.getBytes());
        } catch (IOException e) {
            return ResponseEntity.internalServerError()
                    .body(Map.of("message", "Fehler beim Speichern / Gagal menyimpan"));
        }

        return ResponseEntity.ok(Map.of("message", "Audio gespeichert / Audio disimpan", "fileName", fileName));
    }
}
